core/src/org/javarosa/xpath/expr/XPathPathExpr.java

```diff
@@ -214,59 +214,15 @@ public XPathNodeset eval(DataInstance m, EvaluationContext ec) {
         }
         //Otherwise we'll leave 'm' as set to the main instance 
 
-        //TODO: This causes problems when the paths are heterogeneous. IE: If the path is looking for an attribute that 
-        //doesn't exist on the first node, there is no template path
-
-        // Error out if main DataInstance doesn't match template path of reference.
-        // Cache the check, so it is only performed once.
-        if (ref.isAbsolute()) {
-            if (cachedTemplate == null) {
-                cachedTemplate = m.getTemplatePath(ref);
-            }
-            if (cachedTemplate == null) {
-                return XPathNodeset.ConstructInvalidPathNodeset(ref.toString(), genericRef.toString());
-            }
+        // Error out if there doesn't exist a (template) path along the
+        // reference starting at the main DataInstance.
+        if (ref.isAbsolute() && !m.hasTemplatePath(ref)) {
```

Comment:
Removed strange unused cacheTemplate var.
Please check that my new logic mirrors the old logic